luaui/Widgets/gui_unit_market.lua

```diff
@@ -0,0 +1,480 @@
+			OfferToSell(unitID)
+		end
+	end
+end
+
+local function FindPlayerIDFromTeamID(teamID)
+    local playerList = spGetPlayerList()
+    for i = 1, #playerList do
+        local playerID = playerList[i]
+        local team = select(6,spGetPlayerInfo(playerID))
+        if team == teamID then
+            return playerID
+        end
+    end
+    return nil
+end
+
+local function getTeamName(teamID)
+    local _, _, _, isAITeam = spGetTeamInfo(teamID)
+    if isAITeam then
+        local _, _, _, aiName = spGetAIInfo(teamID)
+        if aiName then
+            local niceName = spGetGameRulesParam('ainame_' .. teamID)
+            if niceName then
+                return niceName
+            else
+                return aiName
+            end
+        else
+            return "AI Team (" .. tostring(teamID)..")"
+        end
+    else
+        local playerID = FindPlayerIDFromTeamID(teamID)
+        if playerID then
+            local playerName, _ = spGetPlayerInfo(playerID, false)
+            return playerName
+        else
+            return "Unknown Team (" .. tostring(teamID)..")"
+        end
+    end
+end
+
+local function unitSale(unitID, price, msgFromTeamID)
+    local unitDefID = spGetUnitDefID(unitID)
+    if not unitDefID then return end
+    local unitDef = UnitDefs[unitDefID]
+    if not unitDef then return end
+    local name = getTeamName(msgFromTeamID)
+    if price > 0 then
+        unitsForSale[unitID] = price
+        local msg = Spring.I18N('ui.unitMarket.sellingUnit', { name = name, unitName = unitDef.translatedHumanName, price = price })
+        Print(msg)
+    else
+        ClearUnitData(unitID)
+    end
+end
+
+local function unitSold(unitID, price, old_ownerID, msgFromTeamID)
+    ClearUnitData(unitID)
+    local unitDefID = spGetUnitDefID(unitID)
+    if not unitDefID then return end
+    local unitDef = UnitDefs[unitDefID]
+    if not unitDef then return end
+    local old_owner_name = getTeamName(old_ownerID)
+    local new_owner_name = getTeamName(msgFromTeamID)
+    if (old_owner_name and new_owner_name) then
+        local msg = Spring.I18N('ui.unitMarket.unitSold', { oldName = old_owner_name, name = new_owner_name, unitName = unitDef.translatedHumanName, price = price })
+        Print(msg)
+    end
+end
+
+local function unitSaleBroadcast(unitID, price, msgFromTeamID)
+    unitSale(unitID, price, msgFromTeamID)
+end
+
+local function unitSoldBroadcast(unitID, price, old_ownerID, msgFromTeamID)
+    unitSold(unitID, price, old_ownerID, msgFromTeamID)
+end
+
+function widget:Initialize()
+    -- if market is disabled, exit
+    if not unitMarket or unitMarket ~= true then
+        widgetHandler:RemoveWidget() -- not enabled? shutdown
+    end
+	-- TODO, in 1vs1 or if you are alone in a team, unless for debug purposes - widget should auto-shutdown
+    -- TODO, if you restart the widget you will forget who is selling any units, create inline that loops through all units and gets param if they are on sale
```

Comment:
I'm not sure if this todo is still relevant, I reloaded the widget and was still able to buy units put up for sale before reloading it.